File: SymmetricTree_LC101.java
//TC:O(n)
//SC:O(h)

class Solution {
    public boolean isSymmetric(TreeNode root) {
      
        //consider left subtree as tree and right subtree as another tree and perform same tree check
     return   helper(root, root);

    }
    
    private boolean helper(TreeNode leftTree, TreeNode rightTree)

    {
        //comparing if both root null then true
       if(leftTree==null && rightTree==null){
           return true;
       }
        
        //comparing if any one root null then return false
    
    if(leftTree==null || rightTree ==null){
        return false;
    }
    
            //if both root value are same check the leftTrees left with right subtree's left
        if(leftTree.val == rightTree.val){
          boolean left = helper(leftTree.left, rightTree.right);
            //check the leftTrees right with right subtree's left
          boolean right = helper(leftTree.right, rightTree.left);
            
            //return boolean value of both trees
            return left&& right;
        }
    
    return false;
    
}}

File: symmetricTree-LC 101.py
# Time Complexity = O(n)
# Space Complexity = O(h), where h is the height of the tree



# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        self.isSymm = True
        
        if root == None:
            return True
        
        self.helper(root.left, root.right)
        
        return self.isSymm
    
    
    def helper(self, left: TreeNode, right: TreeNode):
        # Base Case
        if left == None and right == None:
            return
        
        
        # If both left and right are null, then it should be true, but if the 1st condition is true, it won't go to the 2nd one, which might result in a wrong answer, hence the above base case if added before
        if (left == None or right == None or left.val != right.val):
            self.isSymm = False
            return                      # Returning when once a mismatch occurs so that it doesn't keep recursing over null nodes that will result in runtime error
            
        self.helper(left.left, right.right)
        self.helper(left.right, right.left)